plugin/plugins/filesystem/filesystem.go

```diff
@@ -0,0 +1,138 @@
+
+	// Plugins is an exported list of plugins that will be loaded by go-ipfs.
+	Plugins = []plugin.Plugin{
+		&FileSystemPlugin{}, //TODO: individually name implementations: &P9{}
+	}
+
+	logger logging.EventLogger
+)
+
+func init() {
+	logger = logging.Logger("plugin/filesystem")
+}
+
+type FileSystemPlugin struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	addr      multiaddr.Multiaddr
+	listener  manet.Listener
+	errorChan chan error
+}
+
+func (*FileSystemPlugin) Name() string {
+	return PluginName
+}
+
+func (*FileSystemPlugin) Version() string {
+	return PluginVersion
+}
+
+func (fs *FileSystemPlugin) Init(env *plugin.Environment) error {
+	logger.Info("Initializing 9P resource server...")
+
+	// stabilise repo path; our template depends on this
+	if !filepath.IsAbs(env.Repo) {
+		absRepo, err := filepath.Abs(env.Repo)
+		if err != nil {
+			return err
+		}
+		env.Repo = absRepo
+	}
+
+	cfg := &Config{}
+	// load config from file or initialize it
+	if env.Config != nil {
+		if err := mapstructure.Decode(env.Config, cfg); err != nil {
+			return err
+		}
+	} else {
+		cfg = defaultConfig()
+	}
+
+	var addrString string
+	// allow environment variable to override config values
+	if envAddr := os.ExpandEnv(EnvAddr); envAddr != "" {
+		addrString = EnvAddr
+	} else {
+		addrString = cfg.Service[defaultService]
+	}
+
+	// expand string templates and initialize listening addr
+	templateRepoPath := env.Repo
+	if strings.HasPrefix(addrString, "/unix") {
+		// prevent template from expanding to double slashed paths like `/unix//home/...`
+		// but allow it to expand to `/unix/C:\Users...`
+		templateRepoPath = strings.TrimPrefix(templateRepoPath, "/")
+	}
+
+	addrString = os.Expand(addrString, configVarMapper(templateRepoPath))
+
+	ma, err := multiaddr.NewMultiaddr(addrString)
+	if err != nil {
+		return err
+	}
+	fs.addr = ma
+
+	logger.Info("9P resource server okay for launch")
+	return nil
+}
+
+func (fs *FileSystemPlugin) Start(core coreiface.CoreAPI) error {
+	logger.Info("Starting 9P resource server...")
+
+	// make sure sockets are not in use already (if we're using them)
+	err := removeUnixSockets(fs.addr)
+	if err != nil {
+		return err
+	}
+
+	fs.ctx, fs.cancel = context.WithCancel(context.Background())
+	fs.errorChan = make(chan error, 1)
+
+	// launch the listener
+	listener, err := manet.Listen(fs.addr)
+	if err != nil {
+		logger.Errorf("9P listen error: %s\n", err)
+		return err
+	}
+	fs.listener = listener
+
+	// construct and launch the 9P resource server
+	opts := []nodeopts.AttachOption{nodeopts.Logger(logging.Logger("9root"))}
+	s := p9.NewServer(fsnodes.RootAttacher(fs.ctx, core, opts...))
+	go func() {
+		fs.errorChan <- s.Serve(manet.NetListener(fs.listener))
+	}()
+
+	logger.Infof("9P service is listening on %s\n", fs.listener.Addr())
+	return nil
+}
+
+func (fs *FileSystemPlugin) Close() error {
+	logger.Info("9P server requested to close")
+	fs.listener.Close()
+	fs.cancel()
+	return <-fs.errorChan
```

Comment:
Maybe, I don't know how we're actually using this. If it's not idempotent, we should document that.

Reply:
You're right.
I made changes here:
https://github.com/ipfs/go-ipfs/pull/6612/commits/8cee086b55721be3675a3f01f7e12e71a4830133#diff-8d243ab8514fa1b77d41df37a81641c9R44
https://github.com/ipfs/go-ipfs/pull/6612/commits/8cee086b55721be3675a3f01f7e12e71a4830133#diff-8d243ab8514fa1b77d41df37a81641c9R137
https://github.com/ipfs/go-ipfs/pull/6612/commits/8cee086b55721be3675a3f01f7e12e71a4830133#diff-8d243ab8514fa1b77d41df37a81641c9R166

However, I also had to add in a bool to signify close intent. I'm wondering if there's a better way to handle this https://github.com/ipfs/go-ipfs/pull/6612/commits/8cee086b55721be3675a3f01f7e12e71a4830133#diff-8d243ab8514fa1b77d41df37a81641c9R146 specifically.
We only want to drop the `Accept` error on `Close`.

Reply:
This should be better, thanks for that. https://github.com/ipfs/go-ipfs/pull/6612/commits/c9faaa7ce9cba9f7c17767d38e638982ec8b5475

I also did this, but I'm not sure if it's an improvement.
Instead of assigning directly to the object and rubbing out the error conditionally, we assign it conditionally. Maybe a little more convoluted than the former.
https://github.com/ipfs/go-ipfs/pull/6612/commits/c7cfd7125abb4b749f21502bc9aee142f982ae81

Not sure.